subsys: fs/nvs: reinit initialize addr to zero

Laczen opened the pull request:
nvs.c: reinit(): initialize addr to zero

Fixes issue #9767

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
subsys: fs/nvs: reinit initialize addr to zero

nvs.c: reinit(): initialize addr to zero

Fixes issue #9767

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

ioannisg commented on subsys/fs/nvs/nvs.c:
```diff
@@ -525,7 +525,7 @@ int nvs_reinit(struct nvs_fs *fs)
 	size_t ate_size, empty_len;
 	struct nvs_ate ate;
-	u32_t addr;
+	u32_t addr = 0;
```
Please, indicate the reason for this initialization.
As far as I see, the addr gets initialized in the for loop below and used in line 554.
I suppose the problem occurs if we never execute the for loop at all. Then, is it the right thing to do to use the value 0?

Laczen replied:
Hi @ioannisg, this initialisation is just done because Coverity CID: 187903 complained about using a uninitialised variable. If the loop is never used this means fs->sector_count is 0. In this case the file system would be non-existing. Setting the addr to zero means it is pointing to the start of the file system. More information about this can be found in issue #9767.

Laczen replied:
Hi @aurel32, thanks the clarification. There are two separate issues here: the first is the false positive from Coverty, the second is protecting against changes in the nvs_fs structure.

For the false positive from Coverty setting the addr=0 is the simplest and best solution, as soon as the loop starts it will get this value anyway.

For the protection against changes in the structure this is a different problem, which as you example many parts of zephyr have. Maybe a better way would be to split up the definition of ``nvs_fs`` in a ``const`` part (offset, sector_size and sector_count) and a ``non-const`` part (the rest).
This way the root cause of the problem has been handled without having to do any checks during execution.

aurel32 replied:
> For the false positive from Coverty setting the addr=0 is the simplest and best solution, as soon as the loop starts it will get this value anyway.

This still introduces a change in the generated code, that's why I would like to know if there is 
a way to do that in an other way. Otherwise I agree that your change is the simplest way to fix that.

> For the protection against changes in the structure this is a different problem, which as you example many parts of zephyr have.

I don't think there is anything to do there. As said before, we do not protect for changes in `struct device`, so I don't see why `struct nvs_fs` should be handled differently.

Laczen committed:
subsys: fs/nvs: Make some nvs_fs variables constant

nvs_fs variables offset, sector_count and sector_size are supposed to
be constant by the nvs subsystem. This patch declares them as const.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

aurel32 reviewed (commented):
I have mixed filling those changes. It indeed prevents to write code that will change the values of those struct member.

On the other hand it means that those fields has to be known at compile time and can't come from some probe or computation done before the call to nvs_init. I guess it's possible to workaround that by declaring the `struct nvs_fs` variable on the stack. Note that it also prevents allocating this struct on the heap.

aurel32 commented:
Issue #9767 has been closed by PR #12057, so I think this PR can now be closed. Correct?

Laczen commented:
@aurel32, correct.
